File: mvvm/src/main/java/com/base/mvvm/net/BaseObserver.java
package com.base.mvvm.net;

import android.app.Activity;
import android.content.Context;

import com.base.mvvm.util.DialogUtil;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * 描述: 对观察者进行封装，实现其中部分方法
 *
 * @author : <a href="mailto:fanhuayong@yinli56.com">Fanhy</a>
 * @version : Ver 1.0
 * @date : 2018-03-21 16:16
 */

public abstract class BaseObserver<T> implements Observer {
    private Activity activity;
    private String msg = "";// 进度对话框信息

    public BaseObserver(Activity activity, String msg) {
        this.activity = activity;
        this.msg = msg;
    }

    @Override
    public void onSubscribe(Disposable d) {
        DialogUtil.getInstance().showDialog(activity, msg);
    }

    @Override
    public void onNext(Object value) {
        callBack((T)value);
    }

    @Override
    public void onError(Throwable e) {
        DialogUtil.getInstance().dismissDialog();
    }

    @Override
    public void onComplete() {
        DialogUtil.getInstance().dismissDialog();
    }

    public abstract void callBack(T t);
}

File: kotlindemo/src/main/java/com/base/demo/net/BaseObserver.kt
package com.base.demo.net

import android.app.Activity
import com.base.demo.util.DialogUtil
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

/**
 * 描述: 对观察者进行封装，实现其中部分方法

 * @author : [Fanhy](mailto:fanhuayong@yinli56.com)
 * *
 * @version : Ver 1.0
 * *
 * @date : 2018-03-21 16:16
 */

abstract class BaseObserver<T>(private val activity: Activity, msg: String) : Observer<Any> {
    private var msg = ""// 进度对话框信息

    init {
        this.msg = msg
    }

    override fun onSubscribe(d: Disposable) {
        DialogUtil.instance.showDialog(activity, msg)
    }

    override fun onNext(value: Any) {
        callBack(value as T)
    }

    override fun onError(e: Throwable) {
        DialogUtil.instance.dismissDialog()
    }

    override fun onComplete() {
        DialogUtil.instance.dismissDialog()
    }

    abstract fun callBack(t: T)
}
